src/_gettsim/shared.py

```diff
@@ -39,6 +42,123 @@ def inner(func):
     return inner
 
 
+TIME_DEPENDENT_FUNCTIONS: dict[str, list[Callable]] = {}
+
+
+def dates_active(
+    start: str = "0001-01-01",
+    end: str = "9999-12-31",
+    change_name: Optional[str] = None,
+) -> Callable:
+    """
+    Parameters
+    ----------
+    start
+        The start date (inclusive) in the format YYYY-MM-DD (part of ISO 8601).
+    end
+        The end date (inclusive) in the format YYYY-MM-DD (part of ISO 8601).
+    change_name
+        The name that should be used as the key for the function in the DAG.
+        If omitted, we use the name of the function as defined.
+
+    Returns
+    -------
+        The function with attributes __info__["dates_active_start"],
+        __info__["dates_active_end"], and __info__["dates_active_dag_key"].
+    """
+
+    _validate_dashed_iso_date(start)
+    _validate_dashed_iso_date(end)
+
+    start_date = date.fromisoformat(start)
+    end_date = date.fromisoformat(end)
+
+    _validate_date_range(start_date, end_date)
+
+    def inner(func: Callable) -> Callable:
+        dag_key = change_name if change_name else func.__name__
+
+        _check_for_conflicts(dag_key, func.__name__, start_date, end_date)
+
+        # Remember data from decorator
+        if not hasattr(func, "__info__"):
+            func.__info__ = {}
+        func.__info__["dates_active_start"] = start_date
+        func.__info__["dates_active_end"] = end_date
+        func.__info__["dates_active_dag_key"] = dag_key
+
+        # Register time-dependent function
+        if dag_key not in TIME_DEPENDENT_FUNCTIONS:
+            TIME_DEPENDENT_FUNCTIONS[dag_key] = []
+        TIME_DEPENDENT_FUNCTIONS[dag_key].append(func)
+
+        return func
+
+    return inner
+
+
+_dashed_iso_date = re.compile(r"\d{4}-\d{2}-\d{2}")
+
+
+def _validate_dashed_iso_date(date_str: str):
+    if not _dashed_iso_date.match(date_str):
+        raise ValueError(  # noqa: TRY003
+            f"Date {date_str} does not match the format YYYY-MM-DD."
+        )
+
+
+def _validate_date_range(start: date, end: date):
+    if start > end:
+        raise ValueError(  # noqa: TRY003
+            f"The start date {start} must be before the end date {end}."
+        )
+
+
+def _check_for_conflicts(dag_key: str, function_name: str, start: date, end: date):
+    if dag_key not in TIME_DEPENDENT_FUNCTIONS:
+        return
+
+    for f in TIME_DEPENDENT_FUNCTIONS[dag_key]:
+        # While testing, the same function might be added to the registry again,
```

Comment:
I am unable to follow this comment. Rather add a sentence or two and be very explicit what is happening. I would also suggest:
- changing the name to `_check_for_conflicts_in_time_dependent_functions`
- turning that long comment into the docstring.

Reply:
> turning that long comment into the docstring.

My rule is to use docstrings to document the inferface of an API element, i.e. what its users need to know, and comments to explain the implementation, i.e. what people working on the API element need to know.

I wouldn't mind changing it to a docstring but currently, I just clarified the comment.